Don't rewrite _map files on the save_principal_key redo

dAdAbird opened the pull request:
We create a new WAL key during the extension init, which happens before the redo. This means that in case of a crash,
pg_tde_save_principal_key_redo was rewriting a WAL _map file and destroying a newly created key.

Since we emit an XLog record after the key was successfully written to the file (the file was created), we can safely assume that we should not change the file if it exists.

dAdAbird committed:
Don't rewrite _map files on the save_principal_key redo

We create a new WAL key during the extension init, which happens before
the redo. This means that in case of a crash,
pg_tde_save_principal_key_redo was rewriting a WAL _map file and destroying
a newly created key.

Since we emit an XLog record after the key was successfully written to
the file (the file was created), we can safely assume that we should
not change the file if it exists.

jeltz commented:
I am a bit uncertain about this since I feel it goes against the spirit of what a redo is supposed to do. Like wouldn't this cause issues when replayed on the replica when we have WAL encryption enabled on it? I need to think on this a bit.

dAdAbird committed:
Add crash recovery tests

Tests to trigger redo routines after the server crash. It mostly checks
invariants when different redo functions might rewrite WAL keys created on the
init stage.

For PG-1539, PG-1541, PG-1468, PG-1413

jeltz commented on contrib/pg_tde/t/013_crash_recovery.pl:
```diff
@@ -0,0 +1,90 @@
+$node->kill9;
+PGTDE::append_to_result_file("-- server start");
+PGTDE::append_to_result_file(
+	"-- check the recovery redo and the fix in PR #217 ('Failed to verify principal key header...' error)"
```
I would prefer not to have a reference to a PR here but rather an explanation for what we are actually testing.

jeltz commented on contrib/pg_tde/t/013_crash_recovery.pl:
```diff
@@ -0,0 +1,90 @@
+
+my $node = PostgreSQL::Test::Cluster->new('main');
+$node->init;
+$node->append_conf('postgresql.conf', "shared_preload_libraries = 'pg_tde'");
```
To do a recovery tests shouldn't you set the checkpoint timeout to like 60 minutes to make it clear? Or is that overkill?

dAdAbird replied:
yes, it makes sense to be sure

jeltz reviewed (changes requested):
I still do not trust that this fix is correct but if it is a blocker for your testing of WAL stuff merge it after adressing the feedback to the tests.

jeltz reviewed (approved):
Looks good now.
